### PR TITLE
Fix compilation with intel 17, part 1

### PR DESCRIPTION
**Pull request by drwells**
We need to fix a few things to get things compiling with intel 17 again. Here, without `if constexpr`, we get internal compiler errors due to failed assertions.

### CODE REVIEW

File: source/base/data_out_base.cc
Comment (bangerth) [inline source/base/data_out_base.cc:1510]: can you make it `const`?

File: source/base/data_out_base.cc
Comment (bangerth) [inline source/base/data_out_base.cc:1448]: would you mind making all of these arguments `const`?

File: source/base/data_out_base.cc
Comment (bangerth) [inline source/base/data_out_base.cc:1435]: Could you add a comment to one of these functions? The argument list doesn't give away what the function actually does, but it doesn't seem to have much to do with ordering cells ;-)

**Review by bangerth — CHANGES_REQUESTED** [review]

**Comment by bangerth** [comment]
This fails with the MS VS compiler:
```
  obj_hp_debug.vcxproj -> D:\a\dealii\dealii\build\source\hp\obj_hp_debug.dir\Debug\obj_hp_debug.lib
D:\a\dealii\dealii\include\deal.II/base/table_indices.h(120,1): error C2872: 'internal': ambiguous symbol [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/grid/reference_cell.h(1802,11): message : could be 'dealii::internal' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\source\base\data_out_base.cc(1431,13): message : or       'dealii::`anonymous-namespace'::internal' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/base/symmetric_tensor.h(152): message : see reference to function template instantiation 'dealii::TableIndices<2>::TableIndices<unsigned int,unsigned int>(const unsigned int,const unsigned int)' being compiled [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/base/symmetric_tensor.h(152): message : see reference to function template instantiation 'dealii::TableIndices<2>::TableIndices<unsigned int,unsigned int>(const unsigned int,const unsigned int)' being compiled [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
  Building Custom Rule D:/a/dealii/dealii/source/integrators/CMakeLists.txt
D:\a\dealii\dealii\include\deal.II/base/table_indices.h(121,1): error C2872: 'internal': ambiguous symbol [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/grid/reference_cell.h(1802,11): message : could be 'dealii::internal' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\source\base\data_out_base.cc(1431,13): message : or       'dealii::`anonymous-namespace'::internal' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/base/array_view.h(401,1): error C2872: 'internal': ambiguous symbol [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/grid/reference_cell.h(1802,11): message : could be 'dealii::internal' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\source\base\data_out_base.cc(1431,13): message : or       'dealii::`anonymous-namespace'::internal' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/base/array_view.h(398): message : while compiling class template member function 'dealii::ArrayView<const unsigned int,dealii::MemorySpace::Host>::ArrayView(const unsigned int *,const size_t)' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/grid/reference_cell.h(689): message : see reference to function template instantiation 'dealii::ArrayView<const unsigned int,dealii::MemorySpace::Host>::ArrayView(const unsigned int *,const size_t)' being compiled [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/grid/reference_cell.h(685): message : see reference to class template instantiation 'dealii::ArrayView<const unsigned int,dealii::MemorySpace::Host>' being compiled [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
  Generating affine_constraints.inst
D:\a\dealii\dealii\include\deal.II/base/template_constraints.h(322,1): error C2872: 'internal': ambiguous symbol [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/grid/reference_cell.h(1802,11): message : could be 'dealii::internal' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\source\base\data_out_base.cc(1431,13): message : or       'dealii::`anonymous-namespace'::internal' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\source\base\data_out_base.cc(4267): message : see reference to class template instantiation 'dealii::ProductType<Number,int>' being compiled [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
          with
          [
              Number=double
          ]
  Generating block_sparse_matrix.inst
D:\a\dealii\dealii\include\deal.II/base/point.h(284): message : while compiling class template member function 'dealii::Point<2,ProductType<Number,EnableIfScalar<OtherNumber>::type>::type> dealii::Point<2,Number>::operator /(const OtherNumber) const' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
          with
          [
              Number=double
          ]
D:\a\dealii\dealii\build\source\base\data_out_base.inst(1600): message : see reference to function template instantiation 'void dealii::DataOutBase::write_eps<2>(const std::vector<dealii::DataOutBase::Patch<2,2>,std::allocator<dealii::DataOutBase::Patch<2,2>>> &,const std::vector<std::string,std::allocator<std::string>> &,const std::vector<std::tuple<unsigned int,unsigned int,std::string,dealii::DataComponentInterpretation::DataComponentInterpretation>,std::allocator<std::tuple<unsigned int,unsigned int,std::string,dealii::DataComponentInterpretation::DataComponentInterpretation>>> &,const dealii::DataOutBase::EpsFlags &,std::ostream &)' being compiled [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/base/template_constraints.h(323,1): error C2872: 'internal': ambiguous symbol [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/grid/reference_cell.h(1802,11): message : could be 'dealii::internal' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
D:\a\dealii\dealii\source\base\data_out_base.cc(1431,13): message : or       'dealii::`anonymous-namespace'::internal' [D:\a\dealii\dealii\build\source\base\obj_base_debug.vcxproj]
  fe.cc
```

**Comment by drwells** [comment]
It does (for reasons that shouldn't be hard to fix), but there is actually a much more siginificant problem related to this PR. 

Here is what the UCD output to `tests/data_out/data_out_base_01.cc` looks like on master:
master
and with this patch (which is identical to the VTK output):
feature
this is because the numbering scheme implemented on `master` for DX only works by coincidence and is wrong for UCD. In particular, the new function `set_node_numbers` sets up the internal node numbers for patches in the deal.II numbering, *not* UCD or DX or anything else. *However*, the code
```cpp
    nodes[GeometryInfo<dim>::ucd_to_deal[0]] = start;
```
assumes the RHS values are numbered with UCD (e.g., that `start` is the global index for local vertex number 0), which is not correct.

Unfortunately fixing this bug causes at least a dozen tests to fail. Do we want to go ahead with this patch this close to the release?

**Comment by drwells** [comment]
I personally vote that we fix it but I won't start going through tests if there are strong objections.

**Comment by drwells** [comment]
I went ahead and fixed the issues with this PR. I'd appreciate it if someone could review this soon so I can get back to the rest of the intel 17 fixes.

**Review by kronbichler — APPROVED** [review]
This looks great to me.

**Comment by kronbichler** [comment]
I agree that we should aim to get this merged soon, but let us wait for at least one more approval.

**Review by masterleinad — APPROVED** [review]

**Comment by drwells** [comment]
I think, based on previous discussions, we are all okay with merging in a situation like this (where the concerns have clearly been addressed) without waiting around for a new review, so I am going to go ahead and merge.

**Pull request merged**
